superset-frontend/packages/superset-ui-core/test/utils/lruCache.test.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { lruCache } from '@superset-ui/core';
+
+describe('simple LRUCache', () => {
+  test('initial LRU', () => {
+    expect(lruCache().capacity).toBe(100);
+    expect(lruCache(10).capacity).toBe(10);
+    expect(lruCache(10).size).toBe(0);
+    expect(() => lruCache(0)).toThrow(Error);
+  });
+
+  test('LRU', () => {
```

Comment:
Thanks for the review. This test case is for entire LRU operations.  the L43 depends on the cache instance. The purpose of this test case is to prevent incorrect keys from being inserted at runtime, rather than compile time.

Reply:
Here's an example:

```
const newCache = () => {
  const cache = lruCache<string>(3);
  cache.set('1', 'a');
  cache.set('2', 'b');
  cache.set('3', 'c');
  cache.set('4', 'd');
  return cache;
};

test('initial LRU', () => {
  expect(lruCache().capacity).toBe(100);
  expect(lruCache(10).capacity).toBe(10);
  expect(lruCache(10).size).toBe(0);
  expect(() => lruCache(0)).toThrow(Error);
});

test('correctly evicts LRU value', () => {
  const cache = newCache();
  expect(cache.size).toBe(3);
  expect(cache.has('1')).toBeFalsy();
  expect(cache.get('1')).toBeUndefined();
});

test('correctly updates LRU value', () => {
  const cache = newCache();
  cache.get('2');
  cache.set('5', 'e');
  expect(cache.has('2')).toBeTruthy();
  expect(cache.has('3')).toBeFalsy();
});

test('throws exception when key is invalid', () => {
  const cache = lruCache<string>(3);
  // @ts-ignore
  expect(() => cache.set(0)).toThrow(TypeError);
});

test('clears the cache', () => {
  const cache = newCache();
  cache.clear();
  expect(cache.size).toBe(0);
  expect(cache.capacity).toBe(3);
});
```

This way, when a test fails, it's easier to identify what parts of the code are failing. Also, you don't have dependencies between different test cases. You can also optimize the initial state for each test if desired.

Reply:
Here's an example:

```
const newCache = () => {
  const cache = lruCache<string>(3);
  cache.set('1', 'a');
  cache.set('2', 'b');
  cache.set('3', 'c');
  return cache;
};

test('initial LRU', () => {
  expect(lruCache().capacity).toBe(100);
  expect(lruCache(10).capacity).toBe(10);
  expect(lruCache(10).size).toBe(0);
  expect(() => lruCache(0)).toThrow(Error);
});

test('correctly evicts LRU value', () => {
  const cache = newCache();
  expect(cache.size).toBe(3);
  cache.set('4', 'd');
  expect(cache.has('1')).toBeFalsy();
  expect(cache.get('1')).toBeUndefined();
});

test('correctly updates LRU value', () => {
  const cache = newCache();
  cache.get('1');
  cache.set('5', 'e');
  expect(cache.has('1')).toBeTruthy();
  expect(cache.has('2')).toBeFalsy();
});

test('throws exception when key is invalid', () => {
  const cache = lruCache<string>(3);
  // @ts-ignore
  expect(() => cache.set(0)).toThrow(TypeError);
});

test('clears the cache', () => {
  const cache = newCache();
  cache.clear();
  expect(cache.size).toBe(0);
  expect(cache.capacity).toBe(3);
});
```

This way, when a test fails, it's easier to identify what parts of the code are broken. It also avoids state dependencies between different test cases. You can reuse an initial state or optimize for each test.

Reply:
This is just a test case, do we really need "modularity"? the original test case has 19 lines and the suggestion from you has 32 lines(removed empty line between test case). 

![image](https://user-images.githubusercontent.com/2016594/180799537-75863a2e-0ba6-4911-a54c-bb7d4af15058.png)

the `newCache` is also not a setup fixture, we can't use it in `initial LRU` and `LRU handle null and undefined`. This is a completely additional abstraction,

Reply:
I think readability and separation of concerns are more important than the number of lines. It's not a single test case because they are different requirements. You could have a bug related to a value eviction but have the requirements for clearing the cache and type checking the keys still working.

Reply:
a modularity test case and all-in-one test case are not equivalent, because the each `get` or `set` method will affect the entire LRU.
For example: 
```javascript
test('LRU operation', () => {
  const cache = lruCache<string>(3);
  cache.set('1', 'a');
  cache.set('2', 'b');
  cache.set('3', 'c');
  cache.set('4', 'd');
  expect(cache.size).toBe(3);
....
```

not equal to

```javascript
const newCache = () => {
  const cache = lruCache<string>(3);
  cache.set('1', 'a');
  cache.set('2', 'b');
  cache.set('3', 'c');
  return cache;
};
test('correctly evicts LRU value', () => {
  const cache = newCache();
  expect(cache.size).toBe(3);
  cache.set('4', 'd');
....
```

Reply:
If you move the expect after the set then they will be.

Reply:
The code that I sent you is just an example. Feel free to improve it. For me, the important part is that the each requirement is in its own test case.

Reply:
@michael-s-molina Sorry, my point was that each call to `get` and `set` has an effect on the entire LRU, for example, the `correctly evicts LRU value` and the `correctly updates LRU value` should operate on a single LRU instance rather than separately. For this argument, we should probably get someone else to take a look at it as well, and I'd be happy to revise it, thank you very much for the review.

Reply:
I see no problem in merging these two tests together since they are highly related 😉